src/hivpy/hiv_status.py

```diff
@@ -416,3 +441,145 @@ def disease_and_diagnosis(disease_col, diagnosis_col, disease_rate, diagnosis_pr
+
+    def update_HIV_diagnosis(self, pop: Population):
+        """
+        Diagnose people that have been tested this time step. The default test type used
+        is Ab, but certain policy options make use of PCR (RNA VL) or Ag/Ab tests.
+        Accuracy depends on test sensitivity, PrEP usage, as well as CD4 count.
+        """
+        # tested population in primary infection
+        primary_pop = pop.get_sub_pop([(col.IN_PRIMARY_INFECTION, op.eq, True),
+                                       (col.LAST_TEST_DATE, op.eq, pop.date)])
+
+        if len(primary_pop) > 0:
+            # primary infection diagnosis outcomes
+            diagnosed = pop.transform_group([pop.get_correct_column(col.PREP_TYPE, dt=0),
+                                             pop.get_correct_column(col.PREP_TYPE, dt=1)],
+                                            self.calc_primary_diag_outcomes, sub_pop=primary_pop)
+            # set outcomes
+            pop.set_present_variable(col.HIV_DIAGNOSED, diagnosed, primary_pop)
+            pop.set_present_variable(col.HIV_DIAGNOSIS_DATE, pop.date,
+                                     sub_pop=pop.apply_bool_mask(diagnosed, primary_pop))
+
+            # some people lost at diagnosis
+            lost = pop.transform_group([col.SEX_WORKER], self.calc_primary_loss_at_diag,
+                                       sub_pop=pop.apply_bool_mask(diagnosed, primary_pop))
+            pop.set_present_variable(col.UNDER_CARE, True,
+                                     sub_pop=pop.apply_bool_mask(diagnosed and not lost, primary_pop))
+
+        # remaining tested general population
+        general_pop = pop.get_sub_pop([(col.IN_PRIMARY_INFECTION, op.eq, False),
+                                       (col.LAST_TEST_DATE, op.eq, pop.date)])
+
+        if len(general_pop) > 0:
+            # general diagnosis outcomes
+            r = rng.uniform(size=len(general_pop))
+            # FIXME: should be affected by test type and injectable prep usage
+            diagnosed = r < self.test_sens_general
+            # set outcomes
+            pop.set_present_variable(col.HIV_DIAGNOSED, diagnosed, general_pop)
+            pop.set_present_variable(col.HIV_DIAGNOSIS_DATE, pop.date,
+                                     sub_pop=pop.apply_bool_mask(diagnosed, general_pop))
+
+            # FIXME: should also include onart_tm1 and may need to be affected by date_most_recent_tb
+            # some people lost at diagnosis
+            lost = pop.transform_group([col.SEX_WORKER, pop.get_correct_column(col.NUM_PARTNERS, dt=1),
+                                        pop.get_correct_column(col.ADC, dt=1),
+                                        pop.get_correct_column(col.TB, dt=1),
+                                        pop.get_correct_column(col.NON_TB_WHO3, dt=1)],
+                                       self.calc_general_loss_at_diag,
+                                       sub_pop=pop.apply_bool_mask(diagnosed, general_pop))
```

Comment:
Revisit order of operations and consider necessity of using values from the last time step.

Reply:
Num partners condition should only be applied if `higher_newp_less_engagement` is true

Reply:
higher_newp_less_engagement is a sampled variable (20% chance true)